sympy/holonomic/holonomic.py

```diff
@@ -674,6 +707,137 @@ def __mul__(self, other):
+        an algebraic function.
+
+        Examples
+        ========
+
+        >>> from sympy.holonomic.holonomic import HolonomicFunction, DiffOperatorAlgebra
+        >>> from sympy.polys.domains import ZZ, QQ
+        >>> from sympy import symbols
+        >>> x = symbols('x')
+        >>> R, Dx = DiffOperatorAlgebra(QQ.old_poly_ring(x),'Dx')
+
+        >>> HolonomicFunction(Dx - 1, x).composition(x**2) # e^(x**2)
+        HolonomicFunction((-2*x) + (1)Dx, x)
+
+        See Also
+        ========
+        From_Hyper
+        """
+        R = self.annihilator.parent
+        a = self.annihilator.order
+        diff = expr.diff()
+        listofpoly = self.annihilator.listofpoly
+        for i, j in enumerate(listofpoly):
+            if isinstance(j, self.annihilator.parent.base.dtype):
+                listofpoly[i] = self.annihilator.parent.base.to_sympy(j)
+        r = listofpoly[a].subs({self.x:expr})
+        subs = [-listofpoly[i].subs({self.x:expr}) / r for i in range (a)]
+        coeffs = [S(0) for i in range(a)] # Dkfa[i] == coeff of (D^i f)(a) in D^k (f(a))
+        coeffs[0] = S(1)
+        system = [coeffs]
+        homogeneous = Matrix([[S(0) for i in range(a)]]).transpose()
+        sol = Matrix([[]])
+        while sol.is_zero:
+            coeffs_next = [p.diff() for p in coeffs]
+            for i in range(a - 1):
+                coeffs_next[i + 1] += (coeffs[i] * diff)
+
+            for i in range(a):
+                coeffs_next[i] += (coeffs[-1] * subs[i] * diff)
+            coeffs = coeffs_next
+
+            # check for linear relations
+            system.append(coeffs)
+            sol_tuple = Matrix(system).transpose().gauss_jordan_solve(homogeneous)
+            sol = sol_tuple[0]
+
+        sol = sol / sol_tuple[1]
+        sol = _normalize(sol[0:], R, negative=False)
+        if self._have_init_cond:
+            return HolonomicFunction(sol, self.x, self.x0, self.y0)
+        return HolonomicFunction(sol, self.x)
+
+
+def From_Hyper(func):
+    """
+    Converts Hypergeometric Function to Holonomic.
+
+    Examples
+    =======
+
+    >>> from sympy.holonomic.holonomic import From_Hyper, DiffOperatorAlgebra
+    >>> from sympy import symbols, hyper, S
+    >>> x = symbols('x')
+    >>> From_Hyper(hyper([], [S(3)/2], x**2/4))
+    HolonomicFunction((-x) + (2)Dx + (x)Dx**2, x), f(1) = sinh(1) , f'(1) = -sinh(1) + cosh(1)
+
+    """
+    a = func.ap
+    b = func.bq
+    z = func.args[2]
+    x = z.atoms(Symbol).pop()
+    R, Dx = DiffOperatorAlgebra(QQ.old_poly_ring(x), 'Dx')
+    r1 = 1
+    for i in range(len(a)):
+        r1 = r1 * (x * Dx + a[i])
+    r2 = Dx
+    for i in range(len(b)):
+        r2 = r2 * (x * Dx + b[i] - 1)
+    sol = r1 - r2
+
+    simp = hyperexpand(func)
+
+    if isinstance(simp, Infinity) or isinstance(simp, NegativeInfinity):
+        return HolonomicFunction(sol, x).composition(z)
+
+    def _find_conditons(simp, x, x0, order):
```

Comment:
To find Initial Conditions for the Holonomic Function. Just substitutes the point x0 in the hypergeometric function. This is the case where the hypergeometric function can be computed symbolically using `hyperexpand`. I will also add for the case when the function is not know symbolically using `evalf()` .